src/anndata/_io/h5ad.py

```diff
@@ -115,6 +102,38 @@ def write_h5ad(
         write_elem(f, "uns", dict(adata.uns), dataset_kwargs=dataset_kwargs)
 
 
+def _write_x(
+    f: h5py.Group,
+    x: XDataType,
+    *,
+    as_dense: Container[str],
+    dataset_kwargs: Mapping[str, Any],
+) -> None:
+    if "X" in as_dense and isinstance(x, CSMatrix | BaseCompressedSparseDataset):
+        write_sparse_as_dense(f, "X", x, dataset_kwargs=dataset_kwargs)
+    elif x is None:
+        f.pop("X", None)
```

Comment:
adding this branch make Ruff complain about too high complexity, so I extracted `_write_x` and `_write_raw`.

The only difference other than that branch should be that I pulled the `not adata.isbacked or adata.filename != filepath` condition in front of the `"X" in as_dense and isinstance(adata.X, CSMatrix | BaseCSDataset)`, which might change what happens if `adata.isbacked` and `X` is to be densified, let’s see.